### DISCUSSION TITLE
Question about webview extension origins and web APIs like localStorage

### DISCUSSION THREAD

**Discussion by shannon** [microsoft/vscode-discussions#1093] [Extension Development QnA] [answered]
It seems that previously it was stated that localStorage inside of a webview would not be feasible due to an unreliable origin:

https://github.com/microsoft/vscode/issues/96843
https://github.com/microsoft/vscode/issues/66207

Thus this stance would also apply to other web APIs that rely on a shared origin such as SharedWorkers, or Web Locks . 

However, this does not appear to be exactly correct, as seen here:

https://github.com/microsoft/vscode/blob/main/src/vs/workbench/contrib/webview/browser/webview.ts#L340
https://github.com/microsoft/vscode/blob/main/src/vs/workbench/contrib/webview/browser/webview.ts#L377

The origin is randomly generated but it is stored and keyed by application, machine and extension id.

I also found the following https://github.com/microsoft/vscode/issues/137866, which seems to indicate that the origins should now be stable but I don't see support for these APIs documented.

In my testing these APIs do appear to be available and fully functional.  The origin is always the same for every custom editor tab that I opened for that specific extension.

I just want to make sure this isn't going to change if I build my extension using those APIs.  I obviously don't expect them to always be the same on every machine, but if they are always the same on the machine until you clear the cache for that machine, it seems to me these APIs should behave exactly as expected.

Is there some other reason why these origins might change in so far as they would make these APIs not useful? 
If not, is this something that should be documented?

**Comment by tamuratak** [comment, marked as answer]
You can refer to:

- [Consistent origin for webviews and webview views](https://code.visualstudio.com/updates/v1_73#_consistent-origin-for-webviews-and-webview-views) 
- https://github.com/microsoft/vscode/issues/132464#issuecomment-1317800583

> The origin is randomly generated for each extension and type of webview. The same origin is used across all instances of the webview.

> Currently both normal [webviews](https://code.visualstudio.com/api/extension-guides/webview) and [webview views](https://github.com/microsoft/vscode-extension-samples/tree/main/webview-view-sample) try to maintain a consistent origin. We plan on adopting this for custom editors and notebook webviews next iteration

**Reply by shannon** [reply]
Oh perfect, thank you. That is very helpful.

I'm also wondering if we know if VSCode takes any measures to automatically clean up storage from old origins that will effectively never be used again if they do happen to change. Or for example if the extension is uninstalled.

**Reply by burkeholland** [reply]
Marking this as the answer for now. If you feel this is an error, please feel free to re-open.
